[JENKINS-27565] Fix threading issues with Nodes and Queue

stephenc opened the pull request:
[JENKINS-27565](https://issues.jenkins-ci.org/browse/JENKINS-27565)

In order to schedule work correctly, the Queue needs to have a consistent view of all nodes that are allocated to Jenkins. Similarly, in order to determine which nodes are idle and can thus be removed, the Nodes and Clouds need to have a consistent view of the Queue.

What this all boils down to is that there needs to be a lock to guard critical access to the list of nodes and the actual state of the queue. For UI display purposes there needs to be a path that returns snapshot views, but the code paths that intend to modify the nodes or assign jobs to nodes needs a consistent state.

Testing these issues require pushing Jenkins in terms of using a large number of nodes that are provisioned and deprovisioned through the cloud api. Failure modes that have been observed:
- Jobs starting on a node that is removed before the job starts
- Two jobs running on the same node at the same time when the node has one and only one executor and these are not flyweight tasks
- Deadlocks and slow UI when attempting to fix these issues with the existing locks (i.e. the intrinsic locks on `Jenkins.instance` and `Jenkins.instance.queue`)

The basic plan for this pull request is to replace the locks used to manage the queue and the list of nodes with a `ReadWriteLock` or similar and proactively produce Snapshot views of the queue and list of nodes so that the UI threads do not even have to acquire the read lock.
# Plan of attack
- [x] Pick a recent LTS to base development against (`1.596`)
- [x] Fork branch off mainline release LTS was based off (not the LTS itself as that will complicate merging to master)
- [x] Assign a version number for this branch (`1.596-threadsafe-1-SNAPSHOT`)
- [x] ...
- [x] Move slaves out of global config.xml into a config file per slave
- [x] lots of hacking
- [x] ...
- [x] Revert the version number change to make the merge back to mainline possible
- [x] Merge with `master` branch
- [x] Remove `[WIP]` from pull request
- [x] Merge to master

stephenc committed:
[FIXED JENKINS-27565] Refactor the Queue and Nodes to use a consistent locking strategy

The test system I set up to verify resolution of customer(s)' issues driving this change, required
additional changes in order to fully resolve the issues at hand. As a result I am bundling these
changes:

- Moves nodes to being store in separate config files outside of the main config file (improves performance) [FIXED JENKINS-27562]
- Makes the Jenkins is loading screen not block on the extensions loading lock [FIXED JENKINS-27563]
- Removes race condition rendering the list of executors [FIXED JENKINS-27564] [FIXED JENKINS-15355]
- Tidy up the locks that were causing deadlocks with the once retention strategy in durable tasks [FIXED JENKINS-27476]
- Remove any requirement from Jenkins Core to lock on the Queue when rendering the Jenkins UI [FIXED-JENKINS-27566]

stephenc commented:
@reviewbybees

stephenc committed:
Ooops, forgot that the setting of nodes needs to happen while the Queue lock is held and the computer list needs updating afterwards

stephenc committed:
fix failing tests (hopefully)

stephenc committed:
Aha! the latest failing test should now be ok

jglick commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -201,7 +202,12 @@ public static String xsDate(Calendar cal) {
     }
-    
+
+    public static boolean isExtensionsAvailable() {
```
`@since`

jglick commented on core/src/main/java/hudson/model/AbstractCIBase.java:
```diff
@@ -137,15 +135,20 @@ private void updateComputer(Node n, Map<String,Computer> byNameMap, Set<Computer
+            @Override
+            public void run() {
+                Map<Node,Computer> computers = getComputerMap();
+                for (Map.Entry<Node, Computer> e : computers.entrySet()) {
```
Oof, this is going to be impossible to review. For changes like this I generally prefer to keep the indentation of existing lines untouched, even if the result looks a little weird; can always add a later PR to reindent only.

jglick commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -176,6 +176,21 @@
+    private transient final List<RuntimeException> terminatedBy = Collections.synchronizedList(new ArrayList
+            <RuntimeException>());
+
+    public void recordTermination() {
```
`@since` (can just use `TODO` as the version string as a placeholder to remind the merger)

jglick replied:
Would like some Javadoc too.

jglick commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -176,6 +176,21 @@
+        try {
+            throw new RuntimeException(String.format("Termination requested by %s", Thread.currentThread()));
+        } catch (RuntimeException e) {
+            terminatedBy.add(e);
```
WTF?? Simplify to

```java
terminatedBy.add(new RuntimeException("Termination requested by " + Thread.currentThread()));
```

stephenc replied:
and then findbugs complains!

jglick commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -176,6 +176,21 @@
+        }
+    }
+
+    public List<RuntimeException> getTerminatedBy() {
```
`@since`, Javadoc

jglick replied:
(or can use `@Restricted(NoExternalUse.class)` if this is not intended to be a public API)

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -430,31 +452,39 @@ public synchronized void load() {
         } catch (IOException e) {
             LOGGER.log(Level.WARNING, "Failed to load the queue file " + getXMLQueueFile(), e);
+        } finally {
+            updateSnapshot();
```
Should this not be at the end of the main part of the `try` block, rather than in the `finally` block? What if it throws a `RuntimeException`?

jglick replied:
(and similarly elsewhere)

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -430,31 +452,39 @@ public synchronized void load() {
+            try {
+                XmlFile queueFile = new XmlFile(XSTREAM, getXMLQueueFile());
+                queueFile.write(state);
+                SaveableListener.fireOnChange(this, queueFile);
```
Prefer to fire these events outside the lock.

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -921,7 +1051,7 @@ public synchronized Item getItem(Task t) {
+    public Item getItem(AbstractProject p) {
         return getItem((Task) p);
     }
      */
```
I guess this can just be deleted now.

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -1105,98 +1277,121 @@ protected synchronized void _withLock(Runnable runnable) {
+            // The executors that are currently waiting for a job to run.
+            Map<Executor, JobOffer> parked = new HashMap<Executor, JobOffer>();
+
+            {// update parked (and identify any pending items who's executor has disappeared)
```
`whose`

jglick commented:
Boy lambdas would be really helpful for readability here.

jglick commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -174,80 +187,96 @@ public void run() {
+        provisioningLock.lock();
+        try {
+            lastSuggestedReview = System.currentTimeMillis();
+            Queue.withLock(new Runnable() {
```
Should document the lock acquisition order (e.g., as Javadoc on `provisioningLock`) lest the next person to edit this introduces a deadlock.

jglick commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -174,80 +187,96 @@ public void run() {
+                            jenkins.addNode(node);
+                            LOGGER.log(Level.INFO,
+                                    "{0} provisioning successfully completed. We have now {1,number,integer} computer"
+                                            + "(s)",
```
Awkward formatting, which makes even [this link](https://github.com/jenkinsci/jenkins/pull/1596/files?w=1) show excessive diff hunks.

jglick commented on core/src/main/java/jenkins/model/Nodes.java:
```diff
@@ -0,0 +1,254 @@
+package jenkins.model;
```
license header

jglick commented on core/src/main/java/hudson/model/AbstractCIBase.java:
```diff
@@ -160,37 +163,40 @@ private void updateComputer(Node n, Map<String,Computer> byNameMap, Set<Computer
      * This method tries to reuse existing {@link Computer} objects
      * so that we won't upset {@link Executor}s running in it.
      */
-    protected void updateComputerList(boolean automaticSlaveLaunch) throws IOException {
```
While we are here, consider fixing `Jenkins.setNumExecutors` to call this. Last I checked you had to follow that up with `j.setNodes(j.getNodes())`, which seems ridiculous.

stephenc replied:
Done

jglick replied:
Can probably update `GetEnvironmentOutsideBuildTest` and `NodeProvisionerTest` and `QueueTest` accordingly.

jglick commented on core/src/main/java/jenkins/model/Nodes.java:
```diff
@@ -0,0 +1,254 @@
+                    toRemove.remove(name);
+                    Nodes.this.nodes.put(name, n);
+                }
+                Nodes.this.nodes.keySet().removeAll(toRemove);
```
Do we not need to delete the config dirs of the removed nodes?

jglick replied:
I guess that is done by `save`?

stephenc replied:
Yep

jglick commented on core/src/main/java/jenkins/model/Nodes.java:
```diff
@@ -0,0 +1,254 @@
+ *
+ * @since 1.FIXME
+ */
+public class Nodes implements Saveable {
```
Seems like this should be `@Restricted(NoExternalUse.class)` unless I am missing something?

jglick commented on core/src/main/resources/lib/hudson/executors.jelly:
```diff
@@ -39,102 +39,121 @@ THE SOFTWARE.
-                      </j:choose>
-                    </div>
-                  </td>
+      <j:if test="${!c.offline or (c.offline and !e.idle)}">
```
Beware that this no-op change will conflict with #1610. Suggest reverting changes to this file and redoing as a minimal diff.

jglick commented on core/src/main/java/jenkins/model/Nodes.java:
```diff
@@ -0,0 +1,254 @@
+     * @return {@code true} if and only if the list of nodes is stored in the legacy location.
+     */
+    public boolean isLegacy() {
+        return !new File(jenkins.getRootDir(), "nodes").isDirectory();
```
Would suggest a `@LocalData` test verifying compatibility with the legacy format.

jglick commented:
Have a bunch of minor comments inline but :+1: on the overall approach.

ndeloof commented on core/src/main/java/hudson/model/AbstractCIBase.java:
```diff
@@ -137,15 +135,20 @@ private void updateComputer(Node n, Map<String,Computer> byNameMap, Set<Computer
-                computer.onRemoved();
-                return;
+    /*package*/ void removeComputer(final Computer computer) {
+        Queue.withLock(new Runnable() {
```
Why not use a lambda ?
Oh sorry, JDK8 not yet approved as jenkins minimal runtime :P

stephenc committed:
If the number of executors changes, then we need to update the computer list

stephenc committed:
Document the terminatedBy methods

stephenc committed:
Let's be extra safe with unlocking the lock...

(yes `} finally { updateSnapshot(); } } finally {` looks ugly but it reduces the amount of indentation changes)

stephenc committed:
Move SaveableListener notification outside of the lock

stephenc committed:
Remove dead code while we are at it

stephenc committed:
Grammar

stephenc committed:
Document locking logic

stephenc committed:
Add license header

stephenc committed:
Let's keep this private for now

stephenc committed:
Clarify the cases where we hand-roll in the optimized subset of a save() and how config files are maintained

stephenc committed:
Minimise diff to reduce merge conflicts with #1610

stephenc committed:
added `@since` tag as requested by @jglick

stephenc committed:
Missing @since and @Restricted

stephenc committed:
Arg! damn you copy and paste

jglick commented:
[today’s diff](https://github.com/jenkinsci/jenkins/compare/2da5fb9...c67eddf) @jtnord

jglick commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -201,7 +202,25 @@ public static String xsDate(Calendar cal) {
+     * hit a big honking great monitor lock that will block until the effective extension list has been determined
+     * (as if a plugin fails to start, all of the failed plugin's extensions and any dependent plugins' extensions
+     * will have to be evicted from the list of extensions. In practical terms this only affects the
+     * "Jenkins is loading" screen, but as that screen uses the generic layouts we provide this utility method
```
Not quite sure what you mean by “the generic layouts”. Do you mean that you are working on UI elements injected into the “Jenkins is loading” screen? But I cannot think which; it should certainly not be showing the queue/executor widgets.

stephenc replied:
layout.jelly, sidebar.jelly

jglick replied:
I still do not see the relevance to this PR. What was changed that started looking for extensions from the loading screen?

stephenc replied:
My test system would go nuts if the web load got blocked at this point as you would have a big pile of request threads all bunching up because they all hit the lock at the same time and thus were released at the same time. Fixing this enabled me to get consistent failure from the test system under randomised UI load. I could take it out of the pull request, but that would have complicated my verification testing that the failure modes were fixed by the overall pull request

jglick replied:
I am not complaining about this addition, and I trust that there was a good reason for it, I just have not figured out what that reason was, so I am curious. Are you introducing a `PageDecorator` or causing an existing decorator to begin acquiring the `Queue` lock for some reason?

stephenc replied:
One of the page decorators that ships in standard Jenkins ends up blocking the jenkins is loading screen. Nothing to do with the queue. only here because it knocked out my nicely balanced web load on enough occasions that I decided to fix the issue once and for all

jglick replied:
OK, so really could have been in a separate PR, but included here for convenience in testing. Thanks for explanation.

jglick commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -809,6 +865,28 @@ public final int countExecutors() {
     /**
+     * Used to render the list of executors.
+     * @return a snapshot of the executor display information
+     * @since 1.FIXME
```
BTW no `@since` needed if `NoExternalUse`.

jglick commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -1413,6 +1498,87 @@ public boolean accept(File dir, String name) {
+        }
+    }
+
+    public static class TerminationRequest extends RuntimeException {
```
This does deserve `@since`.

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -430,45 +452,58 @@ public synchronized void load() {
             }
         } catch (IOException e) {
             LOGGER.log(Level.WARNING, "Failed to load the queue file " + getXMLQueueFile(), e);
+        } finally { updateSnapshot(); } } finally {
```
Java 7 would be handy here, I guess: could define a local `AutoCloseable` _[sic]_ encapsulating both the lock and the snapshot, and invoke it using `try`-with-resources.

jtnord commented on core/src/main/java/hudson/model/Executor.java:
```diff
@@ -37,6 +37,7 @@
 import jenkins.model.CauseOfInterruption.UserInterruption;
 import jenkins.model.InterruptedBuildAction;
 import jenkins.model.Jenkins;
+import net.jcip.annotations.GuardedBy;
```
wrong guarded import

stephenc replied:
I maintain that Jenkins is using the wrong ones as there is no JSR defining the javax ones and thus they are subject to change ;-) (but yeah it's unlikely they will change and thus I'll fix)

jtnord commented on core/src/main/resources/lib/hudson/executors.jelly:
```diff
@@ -1,7 +1,7 @@
 The MIT License
 
-Copyright (c) 2004-2010, Sun Microsystems, Inc., Kohsuke Kawaguchi,
+Copyright (c) 2004-2015, Sun Microsystems, Inc., Kohsuke Kawaguchi,
```
I highly doubt that this is copyright in this range by Sun Microsystems.
If you feel like modifying this then add a new line with some new dates and the new copyright holder.

reviewbybees replied:
Second line lists me :-D

Sent from my iPhone

> On 25 Mar 2015, at 13:08, James Nord <notifications@github.com> wrote:
> 
> In core/src/main/resources/lib/hudson/executors.jelly:
> 
> > @@ -1,7 +1,7 @@
> >  <!--
> >  The MIT License
> >  
> > -Copyright (c) 2004-2010, Sun Microsystems, Inc., Kohsuke Kawaguchi,
> > +Copyright (c) 2004-2015, Sun Microsystems, Inc., Kohsuke Kawaguchi,
> I highly doubt that this is copyright in this range by Sun Microsystems.
> If you feel like modifying this then add a new line with some new dates and the new copyright holder.
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> -- 
> You received this message because you are subscribed to the Google Groups "engineering-code-reviews" group.
> To unsubscribe from this group and stop receiving emails from it, send an email to engineering-code-reviews+unsubscribe@cloudbees.com.
> To post to this group, send email to engineering-code-reviews@cloudbees.com.
> To view this discussion on the web visit https://groups.google.com/a/cloudbees.com/d/msgid/engineering-code-reviews/jenkinsci/jenkins/pull/1596/r27117918%40github.com.

jtnord commented on core/src/main/java/hudson/model/Computer.java:
```diff
@@ -1413,6 +1498,87 @@ public boolean accept(File dir, String name) {
+        public TerminationRequest() {
+        }
+
+        public TerminationRequest(String message) {
```
would be helpful to record the date/time here as well.

stephenc commented:
@oldelvet the reason for the funky indentation is to let us easily merge with #1610

stephenc commented:
@oldelvet after we have the couple of pull requests against this area merged, I'll likely do a reformat pull request to tidy up

stephenc committed:
Use the GuardedBy import that the Jenkins core uses by convention

- We will never mind that this is probably the wrong import to use as Jenkins core is consistently using the
  wrong one (If anyone asks, for a `javax` import to be correct, it should have a JSR that has passed through voting
  and not have a JSR that is still subject to change... OTOH one could argue that if the JSR 305 should ever
  become revived from its current dormant state *and* it is decided to include in the JRE *then* the packages
  would get moved out of `javax` and into `java` so they are not in final form anyway... but that pre-supposes
  that the process of standardization does not modify the annotations and by virtue of being in `javax`
  there will be some special case handling by some classloaders which is why depending on the JSR305 versions
  is, in my view, a bad plan... anyway this is moot... Jenkins core is going with JSR305 as it currently stands
  so we call all go down together!)

stephenc committed:
Add some javadoc tags and record the time of the termination request

stephenc committed:
Remove `jenkins.setNodes(jenkins.getNodes())` style hacks to force the master number of executors to be updated

jtnord commented on core/src/main/java/hudson/model/Executor.java:
```diff
@@ -73,7 +78,9 @@
     private final Queue queue;
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
 
+    @GuardedBy("#lock")
```
this should just be "lock" according to the docs.

stephenc committed:
fix guarded by values

jtnord commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -120,9 +125,14 @@ public void spent() {
     private final Label label;
 
-    @GuardedBy("self")
+    @GuardedBy("#provisioningLock")
```
another set of GuardedBys that appear to be using incorrect syntax.

jtnord commented on core/src/main/java/hudson/slaves/RetentionStrategy.java:
```diff
@@ -33,8 +33,10 @@
 import java.util.Collections;
 import java.util.HashMap;
 import jenkins.model.Jenkins;
+import net.jcip.annotations.GuardedBy;
```
this import does not match what is is use throught the rest of Jenkins core (be it correct or incorrect).

stephenc committed:
More fun with GuardedBy

jtnord commented on core/src/main/java/hudson/slaves/RetentionStrategy.java:
```diff
@@ -54,6 +56,7 @@
      * @return The number of minutes after which the strategy would like to be checked again. The strategy may be
      *         rechecked earlier or later that this!
      */
+    @GuardedBy("hudson.model.Queue#lock")
```
no '#' just a dot according the the documentation.

stephenc committed:
Even more fun with GuardedBy

jtnord commented on core/src/main/java/hudson/slaves/RetentionStrategy.java:
```diff
@@ -257,21 +265,13 @@ public synchronized long check(final SlaveComputer c) {
+                    // we've been idle for long enough
+                    logger.log(Level.INFO, "Disconnecting computer {0} as it has been idle for {1}",
+                            new Object[]{c.getName(), Util.getTimeSpanString(idleMilliseconds)});
+                    c.disconnect(OfflineCause.create(Messages._RetentionStrategy_Demand_OfflineIdle()));
```
is this correct - there is no lock held hear and I thought the check worked on a task, so it could have had something just allocated to it couldn't it?

jtnord commented on core/src/main/java/hudson/slaves/RetentionStrategy.java:
```diff
@@ -257,21 +265,13 @@ public synchronized long check(final SlaveComputer c) {
-                            }
-                        }
-                    });
+                    // we've been idle for long enough
```
github won't let me comment on the correct line - but this is called from Queue.withLock - and is **also** synchronized?

Can the synchronized be dropped? - if its called without the lock from somewhere then this I think this could potentially disconnect a slave that has just had work started on it?

stephenc replied:
So the method is guarded by Queue.lock and as such the only reason to retain the synchronized is if there is mutable state held in the retention strategy itself

stephenc committed:
Update the retention strategies... the JSR 305 annotations do not seem to be spec'd to apply to child methods also... bold poorly spec'd annotations

jtnord commented:
:+1:

stephenc commented:
@jglick did you want to comment on the commits after your +1

stephenc commented:
@kohsuke you promised me a review ;-)

jglick commented:
Checked [latest commits](https://github.com/jenkinsci/jenkins/compare/c67eddf...8264236a07e047c202a4681ff325f0aa9ca7b632), looks good, :+1:

orrc commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -201,7 +202,25 @@ public static String xsDate(Calendar cal) {
+     * @return {@code true} if the extensions lists have been populated.
+     * @since 1.FIXME
+     */
+    public static boolean isExtensionsAvailable() {
```
Out of curiosity (ok, and pedantry), is this method name grammatically incorrect due to some Java style convention?
e.g. boolean getters should always start with **`is`** (even though in this case, there is no `boolean extensionsAvailable;`), or can it be named **`are`**`ExtensionsAvailable()`?

stephenc replied:
The java beans specification states that getters start with `get` unless the return type is `boolean` in which case they should start with `is`. As this method is accessed via EL it has to be named in accordance with the Java Beans specification, i.e. `is` that way in the jelly page we can do `<j:if test="${h.extensionsAvailable}">...</j:if>`

orrc replied:
@stephenc Thanks for the detailed explanation!

kohsuke commented:
I think I get the motivation, and perhaps maybe it's just me but I think this is trading incorrectness with more deadlocks.

Historically, there has been a lot more locks in Jenkins, and over time as a result of various bug reports of dead locks, we've been shifting toward more lockless data structures and looser consistency. IIUC, this change swings it back to the othe direction.

For example, `AbstractCIBase.updateComputerList()` calls `AbstractCIBase.killComputer()` with a lock, which in turns calls `Computer.kill()`, which is overriden in `SlaveComputer.kill()`, which calls `SlaveComputer.closeChannel()`, which fires `ComputerListener.onOffline()`. Who knows what people do in extension points like that. I'm sure there'll be more examples like this if I look for more, but I hope one would illustrate the point.

My suggestion is to do it like the "actor" model. You identify the data structures that need to be mutated "atomically" --- queue, list of nodes, executors, etc. And you use something like `SingleLaneExecutorService` so that any attempts to mutate them will now run asynchronously, then execute the actual mutation code in serialized form.

I think most use of `Queue.withLock()` can be safely replaced with just a scheduling of a task instead of the execution. For example, all the use of `AbstractCIBase.removeComputer()` still work correctly if removeComputer() changes from "synchronously remove a computer" to "schedule a removal of a computer in a future." Ditto for `NodeProvisioner.update()`, `ComputerRetentionWork.doRun()`, and so on.

I think you do something like that in JOC IIRC, and I had a lot of success eliminating an entire class of dead locks in the workflow plugin by using the same pattern.

stephenc commented:
@kohsuke the example you cite already held a lock on that code path. It just held a different lock from Queue. So what I am doing here is reducing the number of locks in play and removing a whole bunch of locks on Queue with the Queue.Snapshot

Yes I considered moving to a single lane executor service - i.e. just like Swing does - but I think the code needs a tidy up first before we can move towards that point

stephenc committed:
Merge remote-tracking branch 'origin/master' into threadsafe-node-queue

jglick commented:
> safely replaced with just a scheduling of a task instead of the execution

Unless this is done with a careful design, it makes testing hell. I had to revert one such commit you did in core (pertaining to `AbstractItem.delete`, IIRC) because it caused all sorts of random test failures that could not be controlled.

> I had a lot of success eliminating an entire class of dead locks in the workflow plugin by using the same pattern

Though only by rather complex use of linked `Future`s, which introduced their own share of bugs, and has made testing trickier.

In that case I think it was unavoidable, just pointing out that this is not an uncontroversial recommendation. It may be simpler and safer to just have one lock (as here), document it clearly, and document that any code it calls into may not acquire other locks.

stephenc commented:
Part of the issue, as I see it, is that there are quite a few cases where code needs to do the equivalent of `SwingUtilities.invokeAndWait()` because the inner block of code is needed to make subsequent decisions on. Switching to a single lane executor means that we now have traded a very obvious bug (i.e. a deadlock) for a very subtle bug (i.e. a crossed set of Future instances where the `.get()` methods are out of order)

The JVM gives good tooling for finding deadlocks. It does not give good tooling for finding crossed futures. Thus for such a critical piece of code, the safer path is to stick with locks and reduce their usage.

Some of the places I would like to reduce the locking are, e.g. `Nodes` can eliminate the lock in the addNode method and possibly eliminate the lock in the removeNode method with some code simplification...

I had started down that path, namely moving the Computer:Node association out of AbstractCIBase and into either Nodes or Queue (for queue you basically let the Queue.maintain thread be responsible for removing 'dead' Computer instances) but that significantly complicated the pull request and made it harder to give something that is manageable in terms of risk to merge in.

So I guess what I am saying is that after we get this merged my next step would be to try to remove the requirement for the AbstractCIBase lock used to manage the Computer:Node association, that should bring us down to one and only one great big lock... i.e. the lock on Queue.... and once we are at that point I think there would then be scope to see about using a single lane executor.

If you are worried about specific callbacks being invoked with the lock held, please comment on them directly and I will see if we can move them out of the lock safely without compromising the fix of race conditions that are driving this pull request

stephenc committed:
Merge remote-tracking branch 'origin/master' into threadsafe-node-queue

stephenc commented:
@kohsuke from my PoV (unless there are test failures) 7558a6a is ready to merge. Did you read my response to your comments?

stephenc committed:
Ensure that listeners do not get notified when a computer is killed from a thread with the Queue lock

- We don't know what those listeners will be doing
- I'm not completely happy with having to do it this way, there are some theoretical paths where we end
  up re-acquiring the Queue lock in Computer.kill() but they should not be the normal path

daniel-beck commented:
@jglick

> I had to revert one such commit you did in core (pertaining to `AbstractItem.delete`, IIRC)

00d27176f3a452e38bd0caafebe78c7bc2fa822f

kohsuke commented:
I'll try one more time to push back before giving up.

I agree that asynchrony makes testing hard, but I don't think it's insurmountable problem. We'll just have to provide some means to wait until the system reaches a steady state.

And I don't think this needs a complex use of linked futures. When I search the usage of `Queue.withLock`, most of them do not need to wait for their outcome. The only exception I see is `Executor` getting its task, which is a well-controlled location in that it doesn't call into random extensions. Where else does it happen?

So more concretely, and for the sake of clarity, my suggestion is to turn `Queue.withLock` to behave like `SwingUtilities.invokeLater`. The one exceptional invocation in `Exeutor` that returns a value can be changed to access internal lock of Queue object via friendship.


To continue my justification, I don't understand when Stephen says it's trading deadlocks to out of order `Future.get()`. To me, what makes the actor model more appealing here (than in the workflow plugin) is because the processing is already largely asynchronous (when you tweak `Node`, `Computer` and `Executor` don't react immediately) and so there's nobody blocking on `Future.get()` (and even if there is, you can certainly wait on Futures in arbitrary order.)

Plus, there's mostly no expectation of strong read-after-write consistency here. For example, cancelling an item in the queue and finding it in the queue immediately afterward cannot be distinguished from cancelling an item as another user was about to schedule it. I think we just mostly need to provide that writes are serialized (if you cancel an item and you schedule another one afterward, you expect queue to process them in that order and not swap them.)

Dead locks are not obvious bugs. It's very easy for a seemingly harmless changes to introduce deadlocks, and it's hugely destructive to the operation of Jenkins. I've run Jenkins instances long enough, and I really hate these deadlocks that force me to restart Jenkins.

JVM may have some toolings to find dead locks, but them all are dynamic analysis tools. We don't run them, and we can't meanninfuly run them anyway unless we bring in plugins to the picture.

jglick commented:
The question for me is not whether this PR represents the ideal threading model for `Queue`, but whether it is likely to introduce deadlocks where there were none before. (Any change to a threading model to going to introduce _some_ regression, but it may or may not be serious, and the current state is clearly buggy.) We can hold this up to do more refactoring at the cost of reviewing everything again, and of course rerunning all the load tests.

stephenc commented:
@kohsuke I see where your desire to push back comes from.

So you have analysed the use of Queue.withLock in Jenkins core and on that basis assume that it could be replaced with an invokeLater style behaviour.

The issue is that there are places outside of Jenkins core where you need to acquire a lock and reason about the queue with that lock held. Most of these places are harder to find as currently they use the very bold

```
synchronized (Jenkins.getInstance().getQueue()) {
   ...
}
```

effective construct... but not with quite so obvious a call to `getQueue()`. That's why I added the `withLock` methods... to allow us to be more explicit for operations that need to reason about actions while the queue is not changing (because either those actions will conflict with the queue's intentions or because those actions need to reason about nodes being idle)

One example is that in JOC we need the slave manifest to be correct after we declare a connection to JOC as established, thus in the setChannel method (which needs a lock) we need to reason about the slaves that we have currently. If we switch to an invokeAndWait call (which is what it would have to be) then we would be executing our code from a thread that does not have the lock on DescriptorImpl and thus cannot see the protected state until setChannel completes, so we would have to refactor that code also to use a single lane executor and we would end up with a pair of Futures (one from the Queue.invokeAndWait and the other from DescriptorImpl.invokeAndWait) which would be isomorphic with the much simpler nested locks.

There are other examples.

My point on deadlocks being obvious is this:

* The use of pairs of `Future` on single lane executors is somewhat isomorphous with pairs of `Lock`s. You can write the same bugs about lock acquisition order using pairs of futures just as you can with pairs of locks. 
* What happens when a "deadlock" occurs? You would see jobs stop executing. Now as an admin I want to find out why the jobs are just sitting there in the queue with idle nodes doing nothing.
    * If it is a "deadlock" because of a pair of Futures on the single lane executor that are mutually waiting for each other, the JVM does not give *any* diagnostics to find that case
    * If it is a "deadlock" because of a pair of Threads that are fighting over two monitors and have not used a consistent acquisition order... oh look `jstack` prints out `Deadlock detected` and the stack trace points immediately to the plugin code. The support-core plugin even captures the deadlock stack trace for you.

    In both cases, to work around the issue I will need to restart my Jenkins master.
* And before you say "oh but we could use a timed get in Future" I can counter with "oh but we can use a `Lock#tryLock` which is more responsive"

So as an admin of a Jenkins instance I would favour the approach whereby a bug in a plugin can be immediately identified as such - "jobs stopped executing on nodes because we found this deadlock" - over an impossible for a regular admin to identify - "jobs stopped executing on nodes and we cannot figure out why, here's a stack trace with 2000 threads"

stephenc committed:
Doeth, this didn't pick up the old ones

jglick commented:
That was my concern exactly: deadlocks using `Lock` or traditional monitors can be diagnosed (e.g., from a `support.zip`) and resolved by Java programmers with only modest experience of threading issues, whereas problems caused by inappropriate usage of future-like objects are liable to go for months without being properly identified.

I also fear that changing a currently synchronous method to work asynchronously will be a noticeable incompatibility. Even if the normal usage from the UI is unaffected, functional tests could being failing at random. Certainly these can be corrected to wait for a newly defined future, or (without referring to new APIs) wait for some higher-level condition, but we have to account for that work.

kohsuke commented:
I talked offline with folks and I'm giving up. So to move us forward :+1:

jglick commented:
:+1: as of 17d6d415e678ce5a06fea4cfd3b826f0e432b496

jglick commented:
(you could consider updating `DisplayExecutor` to include info about `AsynchronousExecution`, but this is not a priority and could be done separately)

daniel-beck commented on core/src/main/java/jenkins/model/Nodes.java:
```diff
@@ -0,0 +1,282 @@
+ * @since 1.FIXME
+ */
+@Restricted(NoExternalUse.class) // for now, we may make it public later
+public class Nodes implements Saveable {
```
Should have at least a warning in the changelog that you lose configuration when downgrading Jenkins. [JENKINS-27700](https://issues.jenkins-ci.org/browse/JENKINS-27700) rightfully complains.

oleg-nenashev commented:
The change probably causes https://issues.jenkins-ci.org/browse/JENKINS-27708

stephenc commented:
@oleg-nenashev actually I think https://issues.jenkins-ci.org/browse/JENKINS-27708 is a bug in TCB that you will see on 1.536-1.606 if running Jenkins either on a single core non-hyperthreading master or on a multi-core master under high CPU load. I have provided my analysis in the ticket

oleg-nenashev commented:
@stephenc, thanks for the analysis. Seems my installations on 1.509.4 have not been affected by the potential issue in the plugin, which I definitely confirm. The plugin's rework is definitely required (and seems I'm supposed to maintain the plugin). When I find some time, I'll sync up with @ydubreuil on this topic.

stephenc commented:
Prior to 1.536 each executor picked up its work assignments individually, so the way TCB worked was viable.

The change in 1.536 so that executor threads are only created *after* the work assignment broke TCB's assumption, but because most jenkins masters are multi-core it is not immediately obvious

1.607 ensures consistent behaviour irrespective of the number of cores or the load on the master and this highlights the bug in TCB post 1.536

HTH

daniel-beck commented:
This may have caused [JENKINS-27871](https://issues.jenkins-ci.org/browse/JENKINS-27871).

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -497,17 +498,18 @@ protected void onModified() throws IOException {
+     *
+     * @since 1.FIXME
      */
-    protected volatile NodeList slaves;
```
Note that this introduces a behavioral incompatibility: `Jenkins.getInstance().save()` no longer saves modifications to `Node`s, such as changes to `NodeProperty` state made by means other than the configuration screen.

jglick commented on core/src/main/java/hudson/model/Executor.java:
```diff
@@ -215,9 +240,64 @@ public void recordCauseOfInterruption(Run<?,?> build, TaskListener listener) {
+        } finally {
+            pw.close();
+        }
+        LOGGER.log(WARNING, writer.toString());
```
Would not `FINE` be better?

jglick commented on core/src/main/java/hudson/model/Executor.java:
```diff
@@ -307,6 +430,7 @@ private void finish1(@CheckForNull Throwable problems) {
     }
 
     private void finish2() {
+        for (RuntimeException e1: owner.getTerminatedBy()) LOGGER.log(Level.WARNING, String.format("%s termination trace", getName()), e1);
```
Would not `FINE` be better?

jglick replied:
#1993

jglick commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -173,81 +183,114 @@ public void run() {
+            // Since adding a node should not (in principle) confuse Queue#maintain (it is only removal of nodes
+            // that causes issues in Queue#maintain) we should be able to remove the need for Queue#lock
+            //
+            // TODO once Nodes#addNode is made lock free, we should be able to remove the requirement for Queue#lock
```
see #5450
